api/src/main/java/io/grpc/TlsChannelCredentials.java

```diff
@@ -34,9 +44,82 @@ public static ChannelCredentials create() {
   }
 
   private final boolean fakeFeature;
+  private final byte[] certificateChain;
+  private final byte[] privateKey;
+  private final String privateKeyPassword;
+  private final List<KeyManager> keyManagers;
+  private final byte[] rootCertificates;
```

Comment:
Just started the review so there might something later that would answer this question: why use `byte[]` instead of `PrivateKey`, `Certificate` etc? AFAICS there is no additional dependency or netty vs okhttp issue.

Reply:
We're already using byte[] on server-side for key/cert. We could convert byte[] to a Certificate pretty easily. Although for Netty, we don't actually want to do that. For Netty, there's an advantage of using byte[] as then we can pass it to boringssl and have boringssl manage the data (which allows some faster code paths). Given the Java objects PrivateKey and Certificate, we could serialize them for Netty, but it is just more of a pain and we'd have to write that code.

PrivateKey however is different. There's no built-in parser for parsing a der-encoded key. It isn't _that_ bad, but requires a base64 parser and some levels of hackery. Netty has that implemented. But we don't have an implementation for OkHttp; we'd need to write it.

Users tend to have bytes and not Certificate or PrivateKey. So by aligning the API to what the user's have we make the Credentials more of a shim and move the more complex conversion logic into the transports.